codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java

```diff
@@ -0,0 +1,403 @@
+
+                            writer.openBlock("switch t.Name.Local {", "}", () -> {
+                                writer.openBlock("case $S:", "", serializedTargetMemberName, () -> {
+                                    writer.write("originalDecoder := decoder");
+                                    writer.write("decoder = smithydecoding.NewXMLNodeDecoder(memberDecoder.Decoder, t)");
+                                    writer.openBlock("for {", "}", () -> {
+                                        writer.write("var mv $P", targetMemberSymbol);
+                                        nestedTarget.accept(getMemberDeserVisitor(targetMember, "mv", FunctionalUtils.alwaysTrue()));
+                                        writer.write("col = append(col, mv)");
+                                    });
+                                });
+                                writer.write("decoder = originalDecoder");
+                            });
+                        });
+                    } else {
+                        target.accept(getMemberDeserVisitor(member, "col", FunctionalUtils.alwaysTrue()));
+                    }
+                    writer.write("sv = append(sv, col)");
+                });
+
+                writer.openBlock("default:", "", () -> {
+                    writer.write("// Do nothing and ignore");
+                });
+            });
+        });
+        writer.write("*v = sv");
+        writer.write("return nil");
+    }
+
+    private CollectionShape getCollectionShape(Shape shape) {
+        if (shape.isListShape()) {
+            return shape.asListShape().get();
+        }
+
+        if (shape.isSetShape()) {
+            return shape.asSetShape().get();
+        }
+
+        return null;
+    }
+
+    // Generates deserializer function for collection shapes with xml flattened trait.
+    public void generateFlattenedCollectionDeserializer(GenerationContext context, CollectionShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        MemberShape member = shape.getMember();
+        Symbol memberSymbol = symbolProvider.toSymbol(member);
+
+        Shape target  = context.getModel().expectShape(member.getTarget());
+
+        writer.openBlock("func $L(v *$P, decoder *smithydecoding.XMLNodeDecoder) error {", "}",
+                getUnwrappedMapDelegateFunctionName(context, shape), symbol, () -> {
+                    // Initialize the value now that the start stub has verified that there's something there.
+                    writer.write("var sv $P", symbol);
+                    writer.openBlock("if *v == nil {", "", () -> {
+                        writer.write("sv = make($P, 0)", symbol);
+                        writer.openBlock("} else {", "}", () -> {
+                            writer.write("sv = *v");
+                        });
+                    });
```

Comment:
This exact code block seems to be used in a lot of places with the same variable names, so it probably deserves to be its own function. Json could probably use a similar treatment alongside making variable names consistent across the collection types. (but don't worry about doing that second part here)